wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala

```diff
@@ -115,59 +145,87 @@ class NetworkWorker extends Actor with ActorLogging {
+    context.stop(self)
+  }
+
+  override def sendPayload[A <: Payload with ServerSide](payload: A)
+    (implicit codec: Codec[A], opCodeProvider: OpCodeProvider[A]): Unit = {
+    log.debug("Sending " + payload)
+    val bits = PacketSerialization.outgoing(payload)(sessionCipher)
 
-  sealed trait RealmSessionEvent
+    sendRaw(bits)
+  }
 
-  sealed class PayloadBearingEvent[A <: Payload with ServerSide](payload: A)
-    (implicit codec: Codec[A], opCodeProvider: OpCodeProvider[A])
-    extends RealmSessionEvent {
+  override def sendRaw(payloadbits: BitVector, opCode: OpCodes.Value): Unit = {
+    val bits = PacketSerialization.outgoing(payloadbits, opCode)(sessionCipher)
+
+    sendRaw(bits)
+  }
+
+  override def sendRaw(bits: BitVector): Unit = {
+    context.parent ! TCPHandler.OutgoingPacket(bits)
+  }
+
+  def setAuthenticated(sessionKey: BigInt): Unit = {
+    log.debug(s"Session key set up: $sessionKey")
+    sessionCipher = Some(new SessionCipher(sessionKey))
+  }
+}
+
+object NetworkWorker {
+
+  sealed class PayloadBearingMessage[A <: Payload with ServerSide](payload: A)
+    (implicit codec: Codec[A], opCodeProvider: OpCodeProvider[A]) {
     def serialize(sessionCipher: Option[SessionCipher]): BitVector = {
       PacketSerialization.outgoing(payload)(sessionCipher)(implicitly, implicitly)
     }
   }
 
-  case class EventOutgoing[A <: Payload with ServerSide](payload: A)
+  case class SendPayload[A <: Payload with ServerSide](payload: A)
     (implicit opCodeProvider: OpCodeProvider[A], codec: Codec[A])
-    extends PayloadBearingEvent[A](payload)
+    extends PayloadBearingMessage[A](payload)
 
-  case class EventOutgoingRaw(bits: BitVector, opCode: OpCodes.Value) extends RealmSessionEvent
+  case class SendRawPayload(payloadBits: BitVector, opCode: OpCodes.Value)
 
-  case class EventOutgoingSplit(headerBits: BitVector, payloadBits: BitVector) extends RealmSessionEvent
+  case class SendRaw(bits: BitVector)
 
-  case class EventHandlerFailure(err: Err) extends RealmSessionEvent
+  case class SendSplit(headerBits: BitVector, payloadBits: BitVector)
 
-  case object EventEmptyHandlerFailure extends RealmSessionEvent
+  case class Terminate(delayed: Boolean)
 
-  case class EventTerminate(delayed: Boolean) extends RealmSessionEvent
+  case class TerminateWithPayload[A <: Payload with ServerSide](payload: A)
+    (implicit codec: Codec[A], opCodeProvider: OpCodeProvider[A]) extends PayloadBearingMessage[A](payload)
 
-  case class EventTerminateWithPayload[A <: Payload with ServerSide](payload: A)
-    (implicit codec: Codec[A], opCodeProvider: OpCodeProvider[A]) extends PayloadBearingEvent[A](payload)
+  case class HandlePacket(header: ClientHeader, payloadBits: BitVector)
 
-  case class EventAuthenticated(sessionKey: BigInt) extends RealmSessionEvent
+}
+
+class NetworkWorkerFactory(implicit realm: RealmContextData) extends SessionActorCompanion {
+  override def props: Props = Props(new NetworkWorker()(realm))
```

Comment:
I think so. I've modified the code a little bit such that the syntax checker does not fail.